packages/application-shell/src/components/navbar/navbar.js

```diff
@@ -247,39 +248,55 @@ export const MenuItemDivider = () => (
 
 MenuItemDivider.displayName = 'MenuItemDivider';
 
-// This component basically just wraps the `<RestrictedByPermissions>`
+// This component wraps the `<ApplicationContext>`, `<RestrictedByPermissions>`
 // and the `<ToggleFeature>` components. However, it's necessary to have it as
 // the `<ToggleFeature>` wrapper should be rendered only if the `featureToggle`
 // prop is defined. This is because `<ToggleFeature>` will not render any
 // children if the flag is missing/not found.
-export const ToggledWithPermissions = props => {
-  const permissionsWrapper =
-    props.permissions.length > 0 ? (
-      <RestrictedByPermissions
-        permissions={props.permissions}
-        // Always check that some of the given permissions match.
-        shouldMatchSomePermissions={true}
-      >
-        {props.children}
-      </RestrictedByPermissions>
-    ) : (
-      props.children
-    );
-  return props.featureToggle ? (
-    <ToggleFeature flag={props.featureToggle}>
-      {permissionsWrapper}
-    </ToggleFeature>
-  ) : (
-    permissionsWrapper
-  );
-};
-ToggledWithPermissions.displayName = 'ToggledWithPermissions';
-ToggledWithPermissions.propTypes = {
+export const RestrictedMenuItem = props => (
+  <ApplicationContext
+    render={({ visibilityOverwrites: configuredVisibilityOverwrites }) => {
+      // NOTE: Custom application are activated/deactivated while their
+      // visibility is not controlled via a visibiility overwrite.
+      if (
+        props.visibilityOverwrites &&
+        props.visibilityOverwrites.every(
+          visibilityOverwrite =>
+            configuredVisibilityOverwrites[visibilityOverwrite] === true
+        )
+      )
+        return null;
```

Comment:
This logic will eventually move into a pkg in another PR depending on what else needs to be hidden by visibility overwrites. It is plural as the main menu aggregates all overwrites of all its submenus and if all are true we hide. Any submenu just has one overwrite (see below). A submenus overwrite can be hardcoded in the config or later be CoC where it's "just" the key which is helpful for e.g. custom apps.